test/Chainweb/Test/Pact/ModuleCacheOnRestart.hs

```diff
@@ -0,0 +1,159 @@
+import Data.Word
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+
+import Test.Tasty.HUnit
+import Test.Tasty
+
+import System.LogLevel
+
+-- pact imports
+
+import Pact.Gas.Table
+import Pact.Types.Logger hiding (Logger)
+import Pact.Types.SPV
+
+-- chainweb imports
+
+import Chainweb.BlockHeader
+import Chainweb.BlockHeader.Genesis
+import Chainweb.BlockHeaderDB hiding (withBlockHeaderDb)
+import Chainweb.ChainId
+import Chainweb.Logger
+import Chainweb.Miner.Pact
+import Chainweb.Pact.Backend.RelationalCheckpointer
+import Chainweb.Pact.Backend.Types
+import Chainweb.Pact.PactService
+import Chainweb.Pact.Types
+import Chainweb.Payload.PayloadStore.Types
+import Chainweb.Time
+import Chainweb.Test.Utils
+import Chainweb.Test.Pact.Utils
+import Chainweb.Version
+
+testVer :: ChainwebVersion
+testVer = FastTimedCPM peterson
+
+testChainId :: ChainId
+testChainId = someChainId testVer
+
+tests :: ScheduledTest
+tests =
+      ScheduledTest label $
+      withMVarResource mempty $ \iom ->
+      withRocksResource $ \rocksIO ->
+      withPayloadDb $ \pdb ->
+      withBlockHeaderDb rocksIO genblock $ \bhdb ->
+      withTemporaryDir $ \dir ->
+      testGroup label
+      [
+        withPact' testVer Quiet pdb bhdb dir defaultReorgLimit (Supply iom) action (testCase "initial-start")
+      , after AllSucceed "initial-start" $
+        withPact' testVer Quiet pdb bhdb dir defaultReorgLimit (Check (iom >>= readMVar)) action (testCase "restart")
+
+      ]
+  where
+    label = "Chainweb.Test.Pact.ModuleCacheOnRestart"
+    genblock = genesisBlockHeader testVer testChainId
+    action = initialPayloadState dummyLogger testVer testChainId
+
+data R a = Supply !(IO (MVar a)) | Check (IO a)
+
+withPact'
+    :: PayloadCas cas
+    => ChainwebVersion
+    -> LogLevel
+    -> IO (PayloadDb cas)
+    -> IO BlockHeaderDb
+    -> IO FilePath
+    -> Word64
+    -> R ModuleCache
+    -> PactServiceM cas a
+    -> (Assertion -> TestTree)
+    -> TestTree
+withPact' version logLevel iopdb iobhdb iodir deepForkLimit r act toTestTree =
+    withResource startPact stopPact $ \iof -> toTestTree $ iof >>= \(_,f) -> f act
+  where
+    startPact = do
+        pdb <- iopdb
+        bhdb <- iobhdb
+        dir <- iodir
+        sqlEnv <- startSqliteDb version cid logger (Just dir) Nothing False
+        return $ (sqlEnv,) $ \ps -> do
+            T2 _ pstate <- initPactService'' version cid logger bhdb pdb sqlEnv deepForkLimit ps
+            case r of
+              Supply iomcache -> do
+                  mcache <- iomcache
+                  modifyMVar_ mcache (const (pure (_psInitCache pstate)))
+              Check ioa -> do
+                  a <- ioa
+                  assertBool "The module cache on restart is not the same" (a == (_psInitCache pstate))
+
+    stopPact (sqlEnv, _) = stopSqliteDb sqlEnv
+    logger = genericLogger logLevel T.putStrLn
+    cid = someChainId version
+
+-- we want a special version of initPactService'
```

Comment:
Do you mind leaving a more explicit note?

Reply:
done!